docs/nodekeeper.md

```diff
@@ -0,0 +1,29 @@
+The `Nodekeeper` controller will flag through the `metricNodeDrainFailed` metric generating alerts to SRE-P team on-call that any worker node continue to unsuccessfully drain in spite of `NodeDrain` strategies.
+
+## Inside the Nodekeeper Controller
+- The `Nodekeeper` controller mechanism starts with setting the controller with [Add](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L33) function which add sets up the controller for the operator manager - setting up a watch on the resource. For the `Nodekeeper` controller the mechanism is setting up a watch on the resource of Node.
+
+- Now as already specified, the `Nodekeeper` controller works only towards the worker nodes so for keeping the master nodes aside, an [IgnoreMasterPredicate](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/ignoremaster_predicate.go) is used which makes sure that the controller only targets worker nodes in it's mechanism.
+The `IgnoreMasterPredicate` works on the basis of cache, so it considers all the nodes at first run and re-reconciles  at the next run and starts ignoring the Master nodes.
+
+- The function [Reconcile](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L94) is the main and most important part of the controller, it starts with creating `UpgradeConfigManager` to check if we are in an upgrading stage by specifically checking the `MachineConfig` through [IsUpgrading](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L109) and also checks the history using [GetHistory](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L114) based on the `UpgradeConfig` and get all the nodes from the `Kube-Client`.
+
+- The controller mechanism checks for if a node is cordoned inside the `Reconcile` function at [IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L130) based on [node.go/IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/367c5754bd0267f3b6dfd782f30f52a3dc94a3c8/pkg/machinery/node.go#L13). The [node.go/IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/367c5754bd0267f3b6dfd782f30f52a3dc94a3c8/pkg/machinery/node.go#L13) function checks for an `Unschedulable` and `Tainted` node especially for `TaintEffectNoSchedule` node and if it's matched then sets the node as cordoned and also saves the timestamp at which the node was found cordoned.
```

Comment:
"node especially" = "node specifically" ?

Reply:
Changes made according to the suggestion.